### DISCUSSION TITLE
做了一个语音版的硬件，兼容原版

### DISCUSSION THREAD

**Discussion by jinsonli** [peng-zhihui/ElectronBot#161] [Show and tell]
我更新了语音版本的硬件。

【新的优雅简单版】
我在传感器板上集成了声卡和麦克风。
声卡不用写程序，可以更简单的集成音频。
不管使用哪个版本的主板，等大体硬件调试好后，只需要替换掉传感器板子就可以了。

【视频中的音频开发版】
 核心功能兼容原版，只不过把SD卡换成了WM8978
为了后期能扩展本地存储的功能，加上了SPI-FLASH。
也就是说这个硬件可以兼容目前的上位机和软件功能，后期也是可以扩展文件系统做本地存储的。
【另外】如果不打算学习CODEC相关的应用，仅想使用语音功能，我会推出一个只修改传感器板的版本。
（在传感器板上加入USB声卡）目前正在画PCB，验证OK后会配合SDK/DEMO一起通知大家。

视频在：
https://www.bilibili.com/video/BV18a411v7JS?spm_id_from=333.999.0.0

暂时的一些资料在这里更新， 后期打通和原版 USB上层通讯协议后，可能会合入主线。
https://github.com/jinsonli/ElectronBot-Voice

视频中只是展示了一些最终的效果，实际上实现语音功能有这么几个问题和难点。
1：语音的采样率一般为 44K， 16位，这个也是要加到高速USB的传输协议里去的。需要开发一个专门的USB传输类来 
完成多种传感器数据接收和发送。
2.如果想减轻USB传输带宽的压力，可以用16K采样率的音频。科大讯飞里语音识别里就是16K音频。
甚至，可以仅仅用单向的语音发声，麦克风走USB-HUB那一路，我搜了淘宝有一个很微型的USB麦克风，我打算买来试试看。
3: 在SDK层面，我实现了一个“非阻塞”的音频和视频播放接口。也就是说，在程序的任意时刻调用sdk.play("xxx.wav")，
既可以发送播放语音的消息，然后后台程序自动完成音频的发送。视频也是一样。
这样的好处是，上层应用可以更加方便地执行机器人行为。在任意时刻触发音频/视频/姿态指令信号即可。
目前软件还没有完善，后面会逐步更新。

**Comment by jinsonli** [comment]
@peng-zhihui   大佬，我在视频的最后引用了你的demo视频哈 量子纠缠太魔性了哈哈

**Reply by wenFrankieeee** [reply]
NB,厉害， 不过，真要请教一个问题，我们unity上位机无法驱动舵机（无法使用，也无摄像头画面）；单独用你得舵机调试软件却可以，问题出在哪里呢，求解啊，谢谢

**Reply by jinsonli** [reply]
> NB,厉害， 不过，真要请教一个问题，我们unity上位机无法驱动舵机（无法使用，也无摄像头画面）；单独用你得舵机调试软件却可以，问题出在哪里呢，求解啊，谢谢

我试试看

**Comment by peng-zhihui** [comment]
非常棒，目前看到最好的复刻了，已帮你加到README项目描述~

**Reply by jinsonli** [reply]
感谢大佬的项目！争取继续完善。 @_@

**Comment by peng-zhihui** [comment]
实际上语音功能可以直接集成一个USB声卡芯片的，通过电子的USB-HUB透传到电脑，这样可以大幅降低硬件和固件复杂度，而且音频效果也会更好（HUB只用了3个口，刚好还剩一个）。

**Reply by peng-zhihui** [reply]
另外语音功能可以参考 https://github.com/wzpan/wukong-robot 这个项目，我之前做过一个Pico语音助手就是基于这个的效果还不错，扩展性也很强 https://zhuanlan.zhihu.com/p/72896282

**Reply by jinsonli** [reply]
> 实际上语音功能可以直接集成一个USB声卡芯片的，通过电子的USB-HUB透传到电脑，这样可以大幅降低硬件和固件复杂度，而且音频效果也会更好（HUB只用了3个口，刚好还剩一个）。

那挺好的！ 我试试看！估计得在扩展板上下手了 -- ！

**Reply by jinsonli** [reply]
如果有打算学习音频开发的同学可以参考目前的这个硬件。
WM8978是很不错的一款CODEC，目前国内也出了不少替换的硬件。比如ES8388之类的。

**Reply by jinsonli** [reply]
目前的进展是  PCM2912A的电路板已经出来了。  集成在传感器板上。
测试了播放音频和录音，可以正常运行。
目前问题是有点杂音还在查问题。 等硬件OK后我会把DEMO和 SDK上传。
这个版本有望开源给大家， USB的声卡集成麦克风和小喇叭，可以更简单的实现音频功能。
![微信图片_20220424001654](https://user-images.githubusercontent.com/26651232/164914418-79311c20-a668-4381-84ac-28ae9d6a06c5.jpg)

**Reply by jinsonli** [reply]
新的传感器板子昨晚肝出来了，发出去打板。
这回是正儿八经布线的板子，应该没啥问题，板子回来估计得等节后。
希望没问题拜托。 
![第二版-待验证](https://user-images.githubusercontent.com/26651232/165872194-b874a69d-ca7f-4e99-993a-622df5bcb793.png)

**Reply by coolflyreg** [reply]
挺好，打算把你这个改改，然后把语音部分单独搞一块小板子分离出来，然后通过线接USB

**Reply by jinsonli** [reply]
> 挺好，打算把你这个改改，然后把语音部分单独搞一块小板子分离出来，然后通过线接USB

为啥要单独搞一个语音板子额。 我改的这个传感器板子就已经集成了 语音和录音了。 等测试没问题我就把PCB发出来。
今天早上PCB刚发出去。

**Reply by jinsonli** [reply]
我这里验证没啥问题，音频功能正常，麦克风正常，两个I2C传感器寻址正常，舵机通讯正常，大家有疑问一起讨论~
![e2c2ba05e23165bcbe20ac3524f9be3](https://user-images.githubusercontent.com/26651232/166955840-55d47b1d-9c17-4395-a648-5e77331ce9e1.png)

**Reply by killadm** [reply]
中间不留空的话摄像头会不会装不上？
![QQ截图20220507225501](https://user-images.githubusercontent.com/17094197/167259726-b1e551e2-e65f-4a92-8ac0-cf26412135a8.jpg)

**Reply by jinsonli** [reply]
角度不用仰这么高就行了。传感器板子空间有点小了。 我是这样装的。

**Reply by jinsonli** [reply]
> 
![38447eeb4a54569689d93ea170f024e](https://user-images.githubusercontent.com/26651232/167542169-262e0465-456a-4b56-9bae-4cff296070d7.jpg)
![79dc8445011a656ddf40c4743f32fa4](https://user-images.githubusercontent.com/26651232/167542175-814c49d7-fd82-4694-8de4-e3762928655a.jpg)
![f6e07b7b8aeae77aaae8bf4edd20b8d](https://user-images.githubusercontent.com/26651232/167542181-923ddc35-9ea2-493c-a129-f4489bd06ba0.jpg)

**Reply by GreenShadeZhang** [reply]
这个传感器语音版本用起来有没有bug，没有的话我觉得可以都用这个进行复刻了。

**Comment by maxieahui** [comment]
你好我想问一下，up主用的打印材料是什么呢，感谢感谢，已经一键三连了哈哈哈
